Mathlib/Analysis/SpecialFunctions/MulExpNegSq.lean

```diff
@@ -0,0 +1,193 @@
+noncomputable
+def mulExpNegSq (x : ℝ) := x * exp (- x * x)
+
+theorem mulExpNegSq_apply (x : ℝ) : mulExpNegSq x = x * exp (- x * x) :=  rfl
+
+theorem neg_mulExpNegSq_neg (x : ℝ) : - mulExpNegSq (-x) = mulExpNegSq x  := by
+  simp only [mulExpNegSq, neg_mul, neg_neg, mul_neg]
+
+theorem mulExpNegSq_le_one (x : ℝ) : mulExpNegSq x ≤ 1 := by
+  rw [mulExpNegSq_apply, neg_mul, ← le_div_iff₀ (exp_pos (-(x * x))), exp_neg,
+    div_inv_eq_mul, one_mul]
+  apply le_trans _ (add_one_le_exp (x * x))
+  nlinarith
+
+theorem abs_mulExpNegSq_le_one (x : ℝ) : |mulExpNegSq x| ≤ 1 := by
+  apply abs_le.mpr ⟨_, mulExpNegSq_le_one x⟩
+  rw [← neg_mulExpNegSq_neg, neg_le_neg_iff]
+  exact mulExpNegSq_le_one (-x)
+
+theorem deriv_expNegSq (y : ℝ) :
+    deriv (fun (x : ℝ) => exp (-x * x)) y = -2 * y * exp (-y * y) := by
+  simp only [neg_mul, differentiableAt_neg_iff, differentiableAt_id', DifferentiableAt.mul,
+    _root_.deriv_exp, deriv.neg', deriv_mul, deriv_id'', one_mul, mul_one, neg_add_rev]
+  ring
+
+theorem expNegSq_differentiableAt (y : ℝ) :
+    DifferentiableAt ℝ (fun (x : ℝ) => exp (-x * x)) y := by
+  simp only [neg_mul, differentiableAt_neg_iff, differentiableAt_id', DifferentiableAt.mul,
+    DifferentiableAt.exp, implies_true]
+
+theorem expNegSq_differentiable : Differentiable ℝ (fun (x : ℝ) => exp (-x * x)) :=
+  fun _ => expNegSq_differentiableAt _
+
+theorem deriv_mulExpNegSq (x : ℝ) : deriv (fun x => x * exp (- x * x)) x =
+    exp (-x * x) + x * (-2 * x * exp (-x * x)) := by
+  nth_rw 1 [← one_mul (exp (-x * x)), ← (deriv_id x)]
+  rw [← deriv_expNegSq x]
+  apply deriv_mul differentiableAt_id' (expNegSq_differentiableAt x)
+
+theorem deriv_mulExpNegSq_le_one (x : ℝ) : ‖deriv (fun y ↦ y * exp (-y * y)) x‖₊ ≤ 1 := by
+  rw [← NNReal.coe_le_coe, coe_nnnorm, norm_eq_abs, deriv_mulExpNegSq x]
+  have heq : exp (-x * x) + x * (-2 * x * exp (-x * x)) = exp (-x * x) * (1 + 2 * -x * x) := by
+    ring
+  rw [heq, abs_mul, abs_exp]
+  let y := x * x
+  have hy : y = x * x := by rfl
+  rw [neg_mul, mul_assoc, neg_mul, ← hy]
+  apply mul_le_of_le_inv_mul₀ (zero_le_one' ℝ) (exp_nonneg _)
+  simp [← exp_neg (-y), abs_le]
+  constructor
+  · have twomul_le_exp : 2 * y ≤ exp y := by
+      apply le_trans _ exp_one_mul_le_exp
+      have : 2 ≤ exp 1 := by
+        apply le_of_lt (lt_trans _ exp_one_gt_d9)
+        norm_num
+      apply mul_le_mul_of_nonneg_right this (mul_self_nonneg x)
+    apply le_trans twomul_le_exp _
+    simp only [le_add_iff_nonneg_right, zero_le_one]
+  · apply le_trans (one_le_exp (mul_self_nonneg x)) (le_add_of_nonneg_right _)
+    simp [hy, (mul_self_nonneg x)]
+
+theorem mulExpNegSq_lipschitzWith_one : LipschitzWith 1 mulExpNegSq := by
+  apply lipschitzWith_of_nnnorm_deriv_le
+      (Differentiable.mul differentiable_id' expNegSq_differentiable) deriv_mulExpNegSq_le_one
+
+end mulExpNegSq
+
+section mulExpNegMulSq
+
+/-! ### Definition and properties of `mulExpNegMulSq g ε` -/
+
+variable {E: Type*}
```

Comment:
```suggestion
variable {E : Type*}
```

Reply:
You should add x, epsilon and g to that variable line instead of having them at every lemma.